File: MetaheuristicOptimization/AllMetaheuristic/others/cricket_algorithm.py
import numpy as np
import random
import time
import math
from CoefCalculate import CoefCalculate  # Assuming CoefCalculate is available as a Python module

class CricketAlgorithm:
    def __init__(self, fun, dim, bounds, population_size=25, alpha=0.5, max_iter=1000, tol=1e-6):
        """
        Initialize the Cricket Algorithm optimizer.

        Parameters:
        - fun: Objective function to minimize.
        - dim: Number of dimensions (variables).
        - bounds: Tuple of (lower, upper) bounds for each dimension.
        - population_size: Number of crickets (solutions).
        - alpha: Scaling factor for randomization.
        - max_iter: Maximum number of iterations.
        - tol: Stopping tolerance for convergence.
        """
        self.fun = fun
        self.dim = dim
        self.bounds = np.array(bounds)  # Bounds as [(low, high), ...]
        self.population_size = population_size
        self.alpha = alpha
        self.max_iter = max_iter
        self.tol = tol
        self.beta_min = 0.2
        self.Q_min = 0
        self.solutions = None
        self.fitness = None
        self.best_solution = None
        self.best_fitness = float('inf')
        self.N_iter = 0

    def initialize_solutions(self):
        """Initialize the population of solutions (crickets)."""
        self.solutions = np.random.uniform(self.bounds[:, 0], self.bounds[:, 1], 
                                          (self.population_size, self.dim))
        self.fitness = np.array([self.fun(sol) for sol in self.solutions])
        min_idx = np.argmin(self.fitness)
        self.best_solution = self.solutions[min_idx].copy()
        self.best_fitness = self.fitness[min_idx]

    def simple_bounds(self, s):
        """Apply boundary constraints to a solution."""
        s = np.clip(s, self.bounds[:, 0], self.bounds[:, 1])
        return s

    def alpha_new(self, alpha):
        """Update alpha parameter."""
        delta = 0.97
        return delta * alpha

    def optimize(self):
        """Run the Cricket Algorithm optimization."""
        start_time = time.time()
        self.initialize_solutions()
        Q = np.zeros(self.population_size)
        v = np.zeros((self.population_size, self.dim))
        scale = self.bounds[:, 1] - self.bounds[:, 0]
        history = []  # Store history for plotting

        while self.best_fitness > self.tol and self.N_iter < self.max_iter * self.population_size:
            for i in range(self.population_size):
                # Simulate cricket parameters
                N = np.random.randint(0, 121, self.dim)
                T = 0.891797 * N + 40.0252
                T = np.clip(T, 55, 180)
                C = (5 / 9) * (T - 32)
                V = 20.1 * np.sqrt(273 + C)
                V = np.sqrt(V) / 1000
                Z = self.solutions[i] - self.best_solution

                # Calculate frequency
                F = np.zeros(self.dim)
                for j in range(self.dim):
                    if Z[j] != 0:
                        F[j] = V[j] / Z[j]

                # Compute Q[i] as a scalar (use mean of F to reduce to scalar)
                Q[i] = self.Q_min + np.mean(F - self.Q_min) * np.random.rand()
                v[i] = v[i] + (self.solutions[i] - self.best_solution) * Q[i] + V
                S = self.solutions[i] + v[i]

                # Calculate gamma using CoefCalculate
                SumF = np.mean(F) + 10000  # Ensure scalar
                SumT = np.mean(C)          # Ensure scalar
                gamma = CoefCalculate(SumF, SumT)

                # Update solution based on fitness comparison
                M = np.zeros(self.dim)
                for j in range(self.population_size):
                    if self.fitness[i] < self.fitness[j]:
                        distance = np.sqrt(np.sum((self.solutions[i] - self.solutions[j]) ** 2))
                        PS = self.fitness[i] * (4 * math.pi * (distance ** 2))
                        Lp = PS + 10 * np.log10(1 / (4 * math.pi * (distance ** 2)))
                        Aatm = (7.4 * (np.mean(F) ** 2 * distance) / (50 * (10 ** (-8))))
                        RLP = Lp - Aatm
                        K = RLP * np.exp(-gamma * distance ** 2)
                        beta = K + self.beta_min
                        tmpf = self.alpha * (np.random.rand(self.dim) - 0.5) * scale
                        M = self.solutions[i] * (1 - beta) + self.solutions[j] * beta + tmpf
                    else:
                        M = self.best_solution + 0.01 * np.random.randn(self.dim)

                # Select new solution
                new_solution = S if np.random.rand() > gamma else M
                new_solution = self.simple_bounds(new_solution)

                # Evaluate new solution
                Fnew = self.fun(new_solution)
                self.N_iter += 1

                if Fnew <= self.fitness[i]:
                    self.solutions[i] = new_solution
                    self.fitness[i] = Fnew

                    if Fnew <= self.best_fitness:
                        self.best_solution = new_solution
                        self.best_fitness = Fnew

                self.alpha = self.alpha_new(self.alpha)

            # Store history for this iteration
            history.append((self.N_iter // self.population_size, self.best_solution.copy(), self.best_fitness))
            print(f"Iteration {self.N_iter // self.population_size + 1}: Best Fitness = {self.best_fitness}")

        elapsed_time = time.time() - start_time
        print(f"Number of evaluations: {self.N_iter}")
        print(f"Best solution: {self.best_solution}")
        print(f"Best fitness: {self.best_fitness}")
        print(f"Elapsed time: {elapsed_time:.2f} seconds")

        return self.best_solution, self.best_fitness, history

File: MetaheuristicOptimization/AllMetaheuristic/others/Criket Algorithm_re/cricket_algorithm.m

function [best,fmin,N_iter]=cricket_algorithm(para)
% Default parameters
if nargin<1,  para=[25  0.5];  end
n=para(1);  % Population size, typically 10 to 25
alpha=para(2);
betamin=0.2;
d=2;
pi=3.14;
Lb=-5*ones(1,d);
% Upper limit/bounds/ a vector
Ub=10*ones(1,d);
    %Lower Bond


Qmin=0;         % Frequency minimum
      
% Iteration parameters
tol=10^(-6);    % Stop tolerance
N_iter=0;       % Total number of function evaluations
% Dimension of the search variables


% Initial arrays
Q=zeros(n,1);  
v=zeros(n,d);   

for i=1:n,

 Sol(i,:)=Lb+(Ub-Lb).*rand(1,d);
  Fitness(i)=Fun(Sol(i,:));
  
end

% Find the current best
[fmin,I]=min(Fitness);
best=Sol(I,:);
tic;

% Start the iterations -- Cricket Algorithm
while(fmin>tol)
for i=1:n,
            N(i,:) = randi([0 120],1,d);         
           T(i,:)=0.891797*N(i,:)+40.0252;
           if(T(i,:)>180)
               T(i,:)=180;               
           end
           if(T(i,:)<55)
               T(i,:)=55;
           end
           C(i,:)=(5/9)*(T(i,:)-32);           
           V(i,:)=20.1*sqrt(273+C(i,:));  
           V(i,:)=sqrt(V(i,:))/1000;
           Z(i,:)=(Sol(i,:)-best);             
            if(Z(i,:)==0)
               F(i,:)=0;   
           else
              F(i,:)=V(i,:)/Z(i,:); 
               end
           Q(i,:)=Qmin+(F(i,:)-Qmin)*rand;      
          v(i,:)=v(i,:)+(Sol(i,:)-best)*Q(i)+V(i);
          S(i,:)=Sol(i,:)+v(i,:);          
          SumF=sum(F(i,:))/i+10000;
          SumT=sum(C(i,:))/i;
          gamma=CoefCalculate(SumF,SumT);      
                                
           Solo=Sol;             
           scale=(Ub-Lb);       
           for j=1:n,
               if(Fitness(i)<Fitness(j))
                distance=sqrt(sum((Sol(i,:)-Solo(j,:)).^2));
                PS=Fitness(i)*(4*pi*(distance^2));
                Lp=PS+10*log10(1/4*pi*(distance^2));
                Aatm = (7.4 * ( ((F(i,:)^2)*distance)/ (50*(10^(-8)))));
                RLP=Lp-Aatm;
                K=(RLP)*exp(-gamma*distance.^2);
                beta=K+betamin;
                tmpf=alpha.*(rand(1,d)-0.5).*scale; 
               M(i,:)=Sol(i,:).*(1-beta)+Solo(j,:).*beta+tmpf;
                        
          

          else
             M(i,:)=best+0.01*randn(1,d);   
               end  
          end
          
     % Evaluate new solutions
          if(rand >gamma)
          
              u1(i,:)=S(i,:);
          else
             u1(i,:)=M(i,:);
          end
            u1(i,:)=simplebounds(u1(i,:),Lb,Ub);   
          
           
             Fnew=Fun(u1(i,:));
             
    
           if (Fnew<=Fitness(i)) 
                Sol(i,:)=u1(i,:);
                Fitness(i)=Fnew;
          
          % Update the current best
                                
             if Fnew<=fmin,
                best=u1(i,:);
                fmin=Fnew;

             end
                
           end
           
          
                    alpha=alpha_new(alpha);
        end
                N_iter=N_iter+n;
               
end    
                
    
% Output/display
disp(['Number of evaluations: ',num2str(N_iter)]);
disp(['Best =',num2str(min(best))]);
disp(['fmin: ',num2str(fmin)]);
 
toc;
function [Fonksiyon2]=Fun(u)

  [Fonksiyon2]=Sphere(u);
 
function z= Sphere(u)
n = 2;
s = 0;
for j = 1:n
    s = s+u(j)^2; 
end
z = s;


function alpha=alpha_new(alpha)
delta=0.97;
alpha=delta*alpha;


function s=simplebounds(s,Lb,Ub)
  % Apply the lower bound vector
  ns_tmp=s;
  I=ns_tmp<Lb;
  ns_tmp(I)=Lb(I);
  
  % Apply the upper bound vector 
  J=ns_tmp>Ub;
  ns_tmp(J)=Ub(J);
  % Update this new move 
  s=ns_tmp;
